packages/grid/x-data-grid/src/internals/slotsMigration.ts

```diff
@@ -0,0 +1,40 @@
+// TODO v7: This file exist only to simplify typing between
+// components/componentsProps and slots/slotProps
+// Should be deleted when components/componentsProps are removed
+
+type OptionalKeys<T extends object> = Exclude<
+  { [K in keyof T]: {} extends Pick<T, K> ? K : never }[keyof T],
+  undefined
+>;
+
+type UncapitalizeKeys<T extends object> = Uncapitalize<keyof T & string>;
+
+export type UncapitalizeObjectKeys<T extends object> = {
+  [key in UncapitalizeKeys<Pick<T, OptionalKeys<T>>>]?: Capitalize<key> extends keyof T
+    ? T[Capitalize<key>]
+    : never;
+} & {
+  [key in UncapitalizeKeys<Omit<T, OptionalKeys<T>>>]: Capitalize<key> extends keyof T
+    ? T[Capitalize<key>]
+    : never;
+};
```

Comment:
Not sure if it's a 1:1 replacement, but would this shorter type work?

```tsx
type UncapitalizeObjectKeys<T extends object> = {
  [K in keyof T as (K extends string ? Uncapitalize<K> : K)]: T[K]
}
```
It basically loops over the keys and if the key is a string - it will uncapitalize it, otherwise, it would leave it as is.